BUG: Make sure we don't divide by zero

daschaich opened the pull request:
This should fix the issue I discussed in
https://mail.scipy.org/pipermail/numpy-discussion/2013-July/067076.html

Without the ValueError added here, polyfit can (and does) return negative or nan variances with no warning.

daschaich committed:
Make sure we don't divide by zero

This should fix the issue I discussed in
https://mail.scipy.org/pipermail/numpy-discussion/2013-July/067076.html

Without the ValueError added here, polyfit can (and does) return negative or nan variances with no warning.

charris commented on numpy/lib/polynomial.py:
```diff
@@ -603,6 +603,8 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
         #  (which is implied by a Bayesian uncertainty analysis) includes it.
         #  Plus, it gives a slightly more conservative estimate of uncertainty.
+        if (len(x) - order - 2.0) < 1:
+            raise ValueError("need at least three degrees of freedom for covariance matrix")
```
The message isn't quite right. What is needed is that the number of data points must exceed the degree + 3.

charris commented:
Also needs a test, look in `numpy/lib/tests/test_polynomial.py`.

charris commented:
Could you fix up the error message?

daschaich committed:
Copy in charris's message

Since the number of degrees of freedom is simply the number of data points minus (degree + 1), these two messages are exactly equivalent.  But perhaps "degrees of freedom" is not standard numpy terminology.  Use whatever you prefer; makes no difference to me.

charris commented on numpy/lib/polynomial.py:
```diff
@@ -603,6 +603,8 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
         #  (which is implied by a Bayesian uncertainty analysis) includes it.
         #  Plus, it gives a slightly more conservative estimate of uncertainty.
+        if (len(x) - order - 2.0) < 1:
+            raise ValueError("the number of data points must exceed the degree + 3 for covariance matrix")
```
You need to keep line length < 80 characters. Something like
```
             raise ValueError("the number of data points must exceed "
                              "the degree + 3 for covariance matrix")
```

daschaich committed:
Copy in charris's line break

I'm surprised a string can simply be divided between lines like this; guess I learned something new today.

charris commented:
Looks good. Could you squash the commits into one and rewrite the commit message? You can do that with `git rebase -i HEAD~3` and follow directions. The commit message should begin like your PR title `BUG: Make sure we don't divide by zero in np.polyfit` followed by an explanation. You should also add a test that the error is raised using `assert_raises`. Add the test in `numpy/lib/tests/test_polynomial.py` in the `test_polyfit` functions. after all that, do a force push `git push -f origin HEAD`. BTW, as a rule, never work in master, always work in a branch. Your master is now out of sync with upstream and you will probably want to delete your current local repo after this goes in and start it over.

charris commented:
I updated and squashed in #7747, so closing this. Thanks @daschaich .
